### PR TITLE
Exec: Support run exec in current task

### PR DESCRIPTION
**Pull request by yangyalei**
## Summary
There is a problem when vfork() calls execv() (or execl()) to start a new application:  When the parent thread calls vfork() it receives and gets the pid of the vforked task, and *not*  the pid of the desired execv'ed application.
 see  issue #3334
## Impact
exec series funciton
## Testing
sim

### CODE REVIEW

**Comment by yangyalei** [comment]
@xiaoxiang781216

**Comment by acassis** [comment]
@yangyalei please include the Summary in the commit log message

**Comment by yangyalei** [comment]
> @yangyalei please include the Summary in the commit log message

done

**Review by acassis — APPROVED** [review]

File: binfmt/CMakeLists.txt [hunk 1]
```diff
@@ -61,5 +61,7 @@ if(CONFIG_BUILTIN)
   list(APPEND SRCS builtin.c)
 endif()
 
+include_directories(${CMAKE_SOURCE_DIR}/sched)
```
Comment (anchao) [inline binfmt/CMakeLists.txt:64]: target_include_directories

Comment (yangyalei) [reply binfmt/CMakeLists.txt]: done

File: binfmt/binfmt_execmodule.c [hunk 2]
```diff
@@ -96,6 +97,80 @@ static void exec_ctors(FAR void *arg)
...
+ *
+ * Description:
+ *   Exchange the pid of tasks, and reverse parent-child relationship.
+ *
+ * Input Parameters:
+ *   ptcb  - parent task tcb.
+ *   chtcb - child task tcb.
+ *
+ * Returned Value:
+ *   NULL
```
Comment (pkarashchenko) [inline binfmt/binfmt_execmodule.c:111]:
```suggestion
 *   none
```

File: binfmt/binfmt_execmodule.c [hunk 3]
```diff
@@ -96,6 +97,80 @@ static void exec_ctors(FAR void *arg)
...
+static void exchange_pid(FAR struct tcb_s * ptcb, FAR struct tcb_s * chtcb)
+{
+  pid_t pid;
+  int   pndx;
+  int   chndx;
+
+  DEBUGASSERT(ptcb);
+  DEBUGASSERT(chtcb);
+
+  irqstate_t flags = enter_critical_section();
```
Comment (pkarashchenko) [inline binfmt/binfmt_execmodule.c:124]:
```suggestion
  int   chndx;
  irqstate_t flags;

  DEBUGASSERT(ptcb);
  DEBUGASSERT(chtcb);

  flags = enter_critical_section();
```

File: binfmt/binfmt_execmodule.c [hunk 4]
```diff
@@ -96,6 +97,80 @@ static void exec_ctors(FAR void *arg)
...
+  pid = chtcb->group->tg_pid;
+  chtcb->group->tg_pid = ptcb->group->tg_pid;
+  ptcb->group->tg_pid = pid;
+
+  pid = chtcb->group->tg_ppid;
+  chtcb->group->tg_ppid = ptcb->group->tg_ppid;
+  ptcb->group->tg_ppid = pid;
+
+#ifdef HAVE_GROUP_MEMBERS
+  FAR pid_t *tg_members = chtcb->group->tg_members;
```
Comment (pkarashchenko) [inline binfmt/binfmt_execmodule.c:154]: Move declaration of `tg_members` to the beginning of the function

File: binfmt/binfmt_execmodule.c [hunk 5]
```diff
@@ -96,6 +97,80 @@ static void exec_ctors(FAR void *arg)
...
+  chtcb->group->tg_members = ptcb->group->tg_members;
+  ptcb->group->tg_members = tg_members;
+#endif
+
+#ifdef CONFIG_SCHED_HAVE_PARENT
+#ifdef CONFIG_SCHED_CHILD_STATUS
+  FAR struct child_status_s *tg_children = chtcb->group->tg_children;
```
Comment (pkarashchenko) [inline binfmt/binfmt_execmodule.c:161]: Move declaration of `tg_children` to the beginning of the function

File: binfmt/binfmt_execmodule.c [hunk 6]
```diff
@@ -96,6 +97,80 @@ static void exec_ctors(FAR void *arg)
...
+  chtcb->group->tg_children = ptcb->group->tg_children;
+  ptcb->group->tg_children = tg_children;
+#else
+  uint16_t tg_nchildren = chtcb->group->tg_nchildren;
```
Comment (pkarashchenko) [inline binfmt/binfmt_execmodule.c:165]: Move declaration of `tg_nchildren` to the beginning of the function

File: binfmt/binfmt_execmodule.c [hunk 7]
```diff
@@ -116,7 +191,8 @@ static void exec_ctors(FAR void *arg)
 int exec_module(FAR struct binary_s *binp,
                 FAR const char *filename, FAR char * const *argv,
                 FAR char * const *envp,
-                FAR const posix_spawn_file_actions_t *actions)
+                FAR const posix_spawn_file_actions_t *actions,
+                int exec_inner)
```
Comment (pkarashchenko) [inline binfmt/binfmt_execmodule.c:195]: Why `exec_inner` is `int` and not `bool`?

File: include/nuttx/binfmt/binfmt.h [hunk 8]
```diff
@@ -272,7 +272,8 @@ int unload_module(FAR struct binary_s *bin);
 int exec_module(FAR struct binary_s *binp,
                 FAR const char *filename, FAR char * const *argv,
                 FAR char * const *envp,
-                FAR const posix_spawn_file_actions_t *actions);
+                FAR const posix_spawn_file_actions_t *actions,
+                int exec_inner);
```
Comment (xiaoxiang781216) [inline include/nuttx/binfmt/binfmt.h:276]:
```suggestion
                bool spawn);
```

File: binfmt/binfmt_execmodule.c [hunk 9]
```diff
@@ -292,6 +368,11 @@ int exec_module(FAR struct binary_s *binp,
     }
 #endif
 
+  if (exec_inner)
+    {
+      exchange_pid(this_task(), (FAR struct tcb_s *)tcb);
```
[the remark this answers is not in the document]
Comment (yangyalei) [reply binfmt/binfmt_execmodule.c:384]: It is not suitable to use task_reparent here. We only need to replace the vforked task and spawned task pid. After replacement, the spawned task is the child process of the task who called vfork.

**Comment by acassis** [comment]
Please fix this message:

This branch cannot be rebased safely
Rebasing the commits of this branch on top of the base branch cannot be performed automatically as this would create a different result than a regular merge.

File: binfmt/binfmt_exec.c [hunk 10]
```diff
@@ -36,6 +36,98 @@
 
 #ifndef CONFIG_BINFMT_DISABLE
 
+/****************************************************************************
+ * Internel Functions
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:40]:
```suggestion
 * Private Functions
```

File: binfmt/binfmt_exec.c [hunk 11]
```diff
@@ -36,6 +36,98 @@
...
+      if (attr->stackaddr != NULL)
+        {
+          bin->stackaddr = attr->stackaddr;
+        }
+#endif
+    }
+}
+
+/****************************************************************************
+ * Internel Functions
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:83]: remove

File: binfmt/binfmt_execmodule.c [hunk 12]
```diff
@@ -292,6 +379,11 @@ int exec_module(FAR struct binary_s *binp,
     }
 #endif
 
+  if (spawn)
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_execmodule.c:382]:
```suggestion
  if (!spawn)
```

File: binfmt/binfmt_exec.c [hunk 13]
```diff
@@ -141,25 +215,115 @@ int exec_spawn(FAR const char *filename, FAR char * const *argv,
...
+  leave_critical_section(flags);
+  unload_module(bin);
+errout_with_bin:
+  kmm_free(bin);
+errout:
+  return ret;
+}
+
+/****************************************************************************
+ * Internel Functions
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:237]: remove

File: binfmt/binfmt_exec.c [hunk 14]
```diff
@@ -141,25 +215,115 @@ int exec_spawn(FAR const char *filename, FAR char * const *argv,
...
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: exec_spawn_internel
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:241]: move to private section

File: binfmt/binfmt_exec.c [hunk 15]
```diff
@@ -71,7 +155,8 @@
  *
  ****************************************************************************/
 
-int exec_spawn(FAR const char *filename, FAR char * const *argv,
+static int exec_spawn_internel(
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:158]:
```suggestion
static int exec_internel(FAR const char *filename, FAR char * const *argv,
                         FAR char * const *envp, FAR const struct symtab_s *exports,
                         int nexports, FAR const posix_spawn_file_actions_t *actions,
                         FAR const posix_spawnattr_t *attr, bool spawn)
```

File: binfmt/binfmt_exec.c [hunk 16]
```diff
@@ -100,27 +185,9 @@ int exec_spawn(FAR const char *filename, FAR char * const *argv,
       goto errout_with_bin;
     }
 
-  /* Update the spawn attribute */
+  /* Update the bin spawn attribute */
 
-  if (attr)
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:105]: revert the change

File: binfmt/binfmt_exec.c [hunk 17]
```diff
@@ -132,7 +199,7 @@ int exec_spawn(FAR const char *filename, FAR char * const *argv,
 
   /* Then start the module */
 
-  pid = exec_module(bin, filename, argv, envp, actions);
+  pid = exec_module(bin, filename, argv, envp, actions, false);
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:202]:
```suggestion
  pid = exec_module(bin, filename, argv, envp, actions, spawn);
```

File: binfmt/binfmt_exec.c [hunk 18]
```diff
@@ -141,25 +208,114 @@ int exec_spawn(FAR const char *filename, FAR char * const *argv,
       goto errout_with_lock;
     }
 
-#ifdef CONFIG_BINFMT_LOADABLE
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:144]: revert the change

File: binfmt/binfmt_exec.c [hunk 19]
```diff
@@ -141,25 +208,114 @@ int exec_spawn(FAR const char *filename, FAR char * const *argv,
...
+ *   actions  - The spawn file actions
+ *   attr     - The spawn attributes.
+ *
+ * Returned Value:
+ *   It returns the PID of the exec'ed module.  On failure, it returns
+ *   the negative errno value appropriately.
+ *
+ ****************************************************************************/
+
+int exec_spawn(FAR const char *filename, FAR char * const *argv,
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:264]: call exec_internel directly with spawn==true

File: binfmt/binfmt_exec.c [hunk 20]
```diff
@@ -245,7 +401,8 @@ int exec(FAR const char *filename, FAR char * const *argv,
 {
   int ret;
 
-  ret = exec_spawn(filename, argv, envp, exports, nexports, NULL, NULL);
+  ret = exec_spawn_internel(filename, argv, envp,
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:404]:
```suggestion
  ret = exec_internel(filename, argv, envp,
                      exports, nexports, NULL, NULL, false);
```

File: binfmt/binfmt_exec.c (hunk 19 above)
Comment (yangyalei) [reply binfmt/binfmt_exec.c:264]: add a new spawn param?  I wrote it like this at first, but you suggested add a new function instead of modifying the definition of exec_spawn.
`let's add an internal function to avoid change exec_spawn prototype`

Comment (yangyalei) [reply binfmt/binfmt_exec.c:266]: If I can change exec_spawn prototype, There is no need add the `exec_spawn_internel` function, just delivery the spawn parameter directly to exec_module is fine.
But the shortcomings of this way are also obvious. 1. Have to change the definition of exec_spawn; 2. I need to change all places where exec_spawn is called.

File: binfmt/binfmt_execmodule.c [hunk 21]
```diff
@@ -96,6 +97,91 @@ static void exec_ctors(FAR void *arg)
...
+ * Input Parameters:
+ *   ptcb  - parent task tcb.
+ *   chtcb - child task tcb.
+ *
+ * Returned Value:
+ *   none
+ *
+ ****************************************************************************/
+
+static void exchange_pid(FAR struct tcb_s *ptcb, FAR struct tcb_s *chtcb)
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_execmodule.c:115]: exec_exchange or exec_swap

File: binfmt/binfmt_exec.c [hunk 22]
```diff
@@ -255,4 +303,4 @@ int exec(FAR const char *filename, FAR char * const *argv,
   return ret;
 }
 
-#endif /* !CONFIG_BINFMT_DISABLE */
+#endif /* !CONFIG_BINFMT_DISABLE */
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:306]: revert the change

File: binfmt/binfmt_exec.c (hunk 10 above)
Comment (xiaoxiang781216) [reply binfmt/binfmt_exec.c:40]: not fix yet.

File: binfmt/binfmt_exec.c [hunk 23]
```diff
@@ -64,17 +64,20 @@
...
  *
  ****************************************************************************/
 
-int exec_spawn(FAR const char *filename, FAR char * const *argv,
-               FAR char * const *envp, FAR const struct symtab_s *exports,
-               int nexports, FAR const posix_spawn_file_actions_t *actions,
-               FAR const posix_spawnattr_t *attr)
+int exec_internel(FAR const char *filename, FAR char * const *argv,
+                  FAR char * const *envp,
+                  FAR const struct symtab_s *exports,
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:77]:
```suggestion
                  FAR const struct symtab_s *exports, int nexports,
```

File: binfmt/binfmt_exec.c [hunk 24]
```diff
@@ -64,17 +64,20 @@
...
  *   the negative errno value appropriately.
  *
  ****************************************************************************/
 
-int exec_spawn(FAR const char *filename, FAR char * const *argv,
-               FAR char * const *envp, FAR const struct symtab_s *exports,
-               int nexports, FAR const posix_spawn_file_actions_t *actions,
-               FAR const posix_spawnattr_t *attr)
+int exec_internel(FAR const char *filename, FAR char * const *argv,
+                  FAR char * const *envp,
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_exec.c:76]:
```suggestion
                  FAR char * const *argv, FAR char * const *envp,
```

**Review by xiaoxiang781216 — APPROVED** [review]

File: binfmt/binfmt_execmodule.c [hunk 25]
```diff
@@ -96,6 +97,91 @@ static void exec_ctors(FAR void *arg)
...
+  ptcb->group->tg_ppid = pid;
+
+#ifdef HAVE_GROUP_MEMBERS
+  tg_members = chtcb->group->tg_members;
+  chtcb->group->tg_members = ptcb->group->tg_members;
+  ptcb->group->tg_members = tg_members;
+#endif
+
+#ifdef CONFIG_SCHED_HAVE_PARENT
+#ifdef CONFIG_SCHED_CHILD_STATUS
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_execmodule.c:171]: indent

File: binfmt/binfmt_execmodule.c [hunk 26]
```diff
@@ -96,6 +97,91 @@ static void exec_ctors(FAR void *arg)
...
+  tg_children = chtcb->group->tg_children;
+  chtcb->group->tg_children = ptcb->group->tg_children;
+  ptcb->group->tg_children = tg_children;
+#else
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_execmodule.c:175]: ditto

File: binfmt/binfmt_execmodule.c [hunk 27]
```diff
@@ -96,6 +97,91 @@ static void exec_ctors(FAR void *arg)
...
+  tg_nchildren = chtcb->group->tg_nchildren;
+  chtcb->group->tg_nchildren = ptcb->group->tg_nchildren;
+  ptcb->group->tg_nchildren = tg_nchildren;
+#endif
```
Comment (xiaoxiang781216) [inline binfmt/binfmt_execmodule.c:179]: ditto

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@yangyalei  @xiaoxiang781216 
I noticed that assertion now happens with spresense:elf.

```
Hello, World on stdout
Hello, World on stderr
We failed to open "aflav-sautga-ay!" errno is 2

Memory Usage after program executiotestheader:
****************************************************************************
* Executing hello
****************************************************************************

_assert: Current Version: NuttX  12.0.0 0beba027e7 Sep 15 2023 08:25:18 arm
_assert: Assertion failed : at file: group/group_exitinfo.c:85 task: elf_main 0xd00a005
up_dump_register: R0: 2d02fc80 R1: 2d0318c8 R2: 00000000  R3: 00000080
up_dump_register: R4: 2d02fc80 R5: 2d0318c8 R6: 2d02f624  FP: 0d02c8aa
up_dump_register: R8: 00000000 SB: 00000055 SL: 00000080 R11: 00000000
up_dump_register: IP: 2d02f7cc SP: 2d032320 LR: 0d003979  PC: 0d003979
up_dump_register: xPSR: 40000000 BASEPRI: 00000080 CONTROL: 00000004
up_dump_register: EXC_RETURN: 00000000
dump_stack: User Stack:
dump_stack:   base: 0x2d031cc8
dump_stack:   size: 00002016
dump_stack:     sp: 0x2d032320
stack_dump: 0x2d032320: 0d02c8aa 00000055 2d03194c 0d00a005 2d0318c8 2d02fc80 00000000 0d02c8aa
stack_dump: 0x2d032340: 00000055 7474754e 2d030058 2d032358 00000000 2d032ab8 00000006 0d0044a5
stack_dump: 0x2d032360: 2d030c50 2d030c50 2d032ec0 0d004691 2d030c50 0d007bd9 2e323100 00302e30
stack_dump: 0x2d032380: 00000000 0d007b27 2d032d18 62302a64 30616265 37653732 70655320 20353120
stack_dump: 0x2d0323a0: 33323032 3a383020 313a3532 00000038 2d032938 0d004e3f 2d032938 0d004173
stack_dump: 0x2d0323c0: 6d7261f0 2d032900 00000080 0d005119 2d032938 00000003 2d032780 00000080
stack_dump: 0x2d0323e0: 000000e0 0d02bc58 2d032458 00000000 00000000 0d007a43 00000080 0d00ed99
stack_dump: 0x2d032400: 2d02f7f0 2d032780 00000003 0d00d55f 00000000 00000000 00000000 00000023
stack_dump: 0x2d032420: 0d02bc58 0d014c0b 0d014bf7 0d014b68 00000000 0d00d587 00000023 00000000
stack_dump: 0x2d032440: 00000000 00000000 2d02f198 0d00a10b 00000023 0d014c0b 0d02bc58 00000000
stack_dump: 0x2d032460: 0014f300 000000b8 00000200 0d014c58 00001824 0014db88 0d00a005 00000001
stack_dump: 0x2d032480: 2d031cb0 00000000 00000000 0d007d77 00000000 0d00518f 00000000 00000000
```

**Comment by masayuki2009** [comment]
>@yangyalei @xiaoxiang781216
>I noticed that assertion now happens with spresense:elf (CONFIG_DEBUG_ASSERTIONS=y)

You can reproduce the same assertion with sim:elf (CONFIG_DEBUG_ASSERTIONS=y)

**Comment by yangyalei** [comment]
> > @yangyalei @xiaoxiang781216
> > I noticed that assertion now happens with spresense:elf (CONFIG_DEBUG_ASSERTIONS=y)
> 
> You can reproduce the same assertion with sim:elf (CONFIG_DEBUG_ASSERTIONS=y)

Is there anything special about the hello application? I will reproduce it.

**Comment by masayuki2009** [comment]
> Is there anything special about the hello application?

Nothing special. I just enabled CONFIG_DEBUG_ASSERTIONS for the test.

**Comment by yangyalei** [comment]
> > Is there anything special about the hello application?
> 
> Nothing special. I just enabled CONFIG_DEBUG_ASSERTIONS for the test.

Looks link spresense:elf target need arm environment. I compiled but can't run the target.
Instead, I test in sim:elf with CONFIG_DEBUG_ASSERTIONS=y, It works well.

Can you help to reproduce in sim:elf?

**Comment by masayuki2009** [comment]
>Instead, I test in sim:elf with CONFIG_DEBUG_ASSERTIONS=y, It works well.
>
>Can you show more details to help reproduce in sim:elf?


```
$ git diff
diff --git a/boards/sim/sim/sim/configs/elf/defconfig b/boards/sim/sim/sim/configs/elf/defconfig
index 999b22a726..6916613390 100644
--- a/boards/sim/sim/sim/configs/elf/defconfig
+++ b/boards/sim/sim/sim/configs/elf/defconfig
@@ -17,6 +17,8 @@ CONFIG_BOARDCTL_ROMDISK=y
 CONFIG_BOARD_LOOPSPERMSEC=0
 CONFIG_BOOT_RUNFROMEXTSRAM=y
 CONFIG_BUILTIN=y
+CONFIG_DEBUG_ASSERTIONS=y
+CONFIG_DEBUG_FEATURES=y
 CONFIG_DEBUG_SYMBOLS=y
 CONFIG_DEV_LOOP=y
 CONFIG_DEV_ZERO=y
@@ -39,6 +41,7 @@ CONFIG_INIT_ENTRYPOINT="nsh_main"
 CONFIG_LIBC_ENVPATH=y
 CONFIG_LIBC_MAX_EXITFUNS=1
 CONFIG_LIBC_STRERROR=y
+CONFIG_NDEBUG=y
 CONFIG_NSH_ARCHINIT=y
 CONFIG_NSH_BUILTIN_APPS=y
 CONFIG_NSH_FILE_APPS=y
...

$ ./nuttx 

NuttShell (NSH) NuttX-12.0.0
MOTD: username=admin password=Administrator
nsh> elf
Initial memory usage: 1332408
elf_main: Registering romdisk at /dev/ram6

Memory Usage after romdisk_register:
  Before:  1332408 After:  1332536 Change:      128
elf_main: Mounting ROMFS filesystem at target=/mnt/romfs with source=/dev/ram6

Memory Usage after mount:
  Before:  1332536 After:  1333272 Change:      736
testheader: 
****************************************************************************
* Executing errno
****************************************************************************


Memory Usage after exec:
  Before:  1333272 After:  1403160 Change:    69888
elf_main: Wait a bit for test completion
Hello, World on stderr
We failed to open "aflav-sautga-ay!" errno is 2
Hello, World on stdout
nsh> 
Memory Usage after program execution:
  Before:  1403160 After:  1401848 Change:    -1312
testheader: 
****************************************************************************
* Executing hello
****************************************************************************

_assert: Current Version: NuttX  12.0.0 0beba027e7-dirty Sep 15 2023 13:21:27 sim
_assert: Assertion failed : at file: group/group_exitinfo.c:85 task: elf 0x55732de8e247
Getting ready to say "Hello, world"

Hello, world!
It has been said.
...

```

**Comment by masayuki2009** [comment]
@yangyalei @xiaoxiang781216 

Let me revert this PR until you fix the above issue.

**Comment by yangyalei** [comment]
> @yangyalei @xiaoxiang781216
> 
> Let me revert this PR until you fix the above issue.

OK, I have know the root cause and will modify it in the new PR.
